Check if tool is installed before using it

piotr-binkowski opened the pull request:
Currently all tools are used without checking if a particular tool is installed, in #83 it was suggested to test if tools are present before using them and this adds support for checking that.
If tool is not present then testing it will be skipped and it won't be included in the report.

Closes #83

tgorochowik commented on tools/runners/Icarus.py:
```diff
@@ -4,9 +4,10 @@
 class Icarus(BaseRunner):
     def __init__(self):
         super().__init__("icarus")
+        self.ex = 'iverilog'
```
Can we rename `ex` to something self-explanatory?

piotr-binkowski replied:
`executable`?

tgorochowik replied:
+1

A comment about this in the main class would be nice too

tgorochowik replied:
Maybe the executable could also be a kwarg in the main class constructor?

tgorochowik replied:
So you could just call `super().__init__("icarus", "iverilog")` etc

tgorochowik commented on tools/BaseRunner.py:
```diff
@@ -34,3 +35,11 @@ def run(self, tmp_dir, params):
+# This method is called by the main runner script (tools/runner) as a sanity
+# check to verify that tool is properly installed.
+# If False is returned then the test is skipped.
+# @return True when tool is installed and can be used, False otherwise.
```
Why are the comments not a proper python doc strings?

Please convert the comments here (and everywhere) using the proper conventions (see https://www.python.org/dev/peps/pep-0257/ )

tgorochowik commented on tools/runner:
```diff
@@ -54,6 +54,9 @@ new_path = [os.path.abspath(dirs['out'] + "/runners/bin/"), os.environ['PATH']]
 os.environ['PATH'] = ":".join(new_path)
 
+if not runner_obj.can_run():
+    sys.exit(0)
```
How does this resolve #83? 

To actually solve #83 I think that we should exclude the runner from the report completely, not "fake-pass" all the tests.

Additionally I don't think it should be quiet. We should at least show a warning about skipping a runner.

Thirdly I think there still should be a way to force-require all runners (could be a Makefile flag). We don't want to exclude anything in CI by accident.

piotr-binkowski replied:
By doing a "fake-pass" at this stage no log file is generated and thus the column for a particular tool won't be included in the report.

Detecting available tools could be moved to main Makefile and we could handle the warnings and hard requirement for all runners there

tgorochowik replied:
That makes sense. But still - at least a part of it should be moved to makefile (to be able to error on this in CI).

I am also wondering if we could somehow not invoke the runners that fail on can run (am I right to think that this will be executed for each test anyway? We should be able to check it just once).

piotr-binkowski commented:
I've fixed the minor issues and moved the tool checks to makefile 
It is now possible to provide flag `ALL_RUNNERS=1` to makefile that forces it to fail if any of the tools is missing

hzeller commented on tools/BaseRunner.py:
```diff
@@ -34,3 +44,13 @@ def run(self, tmp_dir, params):
+
+        Returns True when tool is installed and can be used, False otherwise.
+        """
+        return True if shutil.which(self.executable) is not None else False
```
Can't this be formulated more readably as
```
  return shutil.which(self.executable) is not None
```
We return a boolean, so not need to convert a boolean to a boolean :)

piotr-binkowski replied:
Thanks for pointing that out, it looks better this way

tgorochowik commented on Makefile:
```diff
@@ -47,9 +49,10 @@ generate-$(1):
-RUNNERS := $(basename $(RUNNERS))
+RUNNERS_FOUND := $(wildcard $(RUNNERS_DIR)/*.py)
+RUNNERS_FOUND := $(RUNNERS_FOUND:$(RUNNERS_DIR)/%=%)
+RUNNERS_FOUND := $(basename $(RUNNERS_FOUND))
```
Can we use this variable when calling the `runner_gen` macro? Otherwise we get lots of "empty runs" I mentioned in my previous comment.

piotr-binkowski replied:
Maybe the variable names are misleading but `RUNNERS` is the variable containing runners that passed the checks and `RUNNERS_FOUND` contains all runners which are present in `tools/runners`.

tgorochowik replied:
Ok, great, thanks!
